control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -285,138 +291,168 @@ impl ResourcePruning for OperationGuardArc<VolumeSnapshot> {
             });
         }
-
         let timestamp = DateTime::<Utc>::from(response.snap_time);
         // What if snapshot succeeds but we can't fetch the replica snapshot, should we carry
         // on as following, or should we bail out?
-        let node = registry.node_wrapper(&replica.node).await?;
-
-        let snapshot = NodeWrapper::fetch_update_snapshot_state(
-            &node,
-            ReplicaSnapshotInfo::new(
-                replica_snap.spec().source_id().replica_id(),
-                replica_snap.spec().uuid().clone(),
-            ),
-        )
-        .await?;
-
-        replica_snap.complete_vol(
-            snapshot.timestamp().into(),
-            snapshot.replica_size(),
-            snapshot.allocated_size() + snapshot.predecessor_alloc_size(),
-        );
-        Ok(VolumeSnapshotCreateResult::new_ok(replica_snap, timestamp))
+        for (replica, replica_snap) in prep_params.replica_snapshot.iter_mut() {
+            let node = registry.node_wrapper(&replica.node).await?;
+            let snapshot = NodeWrapper::fetch_update_snapshot_state(
+                &node,
+                ReplicaSnapshotInfo::new(
+                    replica_snap.spec().source_id().replica_id(),
+                    replica_snap.spec().uuid().clone(),
+                ),
+            )
+            .await?;
+
+            replica_snap.complete_vol(
+                snapshot.timestamp().into(),
+                snapshot.replica_size(),
+                snapshot.allocated_size() + snapshot.predecessor_alloc_size(),
+            );
+        }
+        let snapshots = prep_params
+            .replica_snapshot
+            .iter()
+            .map(|(_, replica_snapshot)| replica_snapshot.clone())
+            .collect::<Vec<_>>();
+        Ok(VolumeSnapshotCreateResult::new_ok(snapshots, timestamp))
     }
+
     async fn snapshot_replica<N: ReplicaSnapshotApi>(
         &self,
-        prep_params: &PrepareVolumeSnapshot,
-        target_node: N,
+        prep_params: &mut PrepareVolumeSnapshot,
+        registry: &Registry,
     ) -> Result<VolumeSnapshotCreateResult, SvcError> {
-        let mut replica_snap = prep_params.replica_snapshot.clone();
         let volume_params = prep_params.parameters.params().clone();
-
-        let replica_params = volume_params.with_uuid(replica_snap.1.spec().uuid());
-        let response = target_node
-            .create_repl_snapshot(&CreateReplicaSnapshot::new(SnapshotParameters::new(
-                replica_snap.1.spec().source_id().replica_id(),
-                replica_params,
-            )))
-            .await?;
-        let timestamp = response.timestamp();
-        replica_snap.1.complete_vol(
-            timestamp.into(),
-            response.replica_size(),
-            response.allocated_size() + response.predecessor_alloc_size(),
-        );
+        let mut timestamp = SystemTime::now();
+
+        for (replica, replica_snap) in prep_params.replica_snapshot.iter_mut() {
```

Comment:
It's a bit confusing why we are modifying this input parameter. From what I can tell we don't use that after as we return via fn output. 
So I suggest instead you clone it beforehand:

```suggestion
        let mut replica_snapshots = prep_params.replica_snapshot.clone();
        for (replica, replica_snap) in replica_snapshots.iter_mut() {
```

Reply:
fixed